src/H5Dint.c

```diff
@@ -3650,19 +3650,22 @@ H5D_get_create_plist(const H5D_t *dset)
             uint8_t *bkg_buf = NULL; /* Background conversion buffer */
             size_t   bkg_size;       /* Size of background buffer */
 
-            if (NULL == (src_type = H5T_copy(dset->shared->type, H5T_COPY_ALL)))
-                HGOTO_ERROR(H5E_DATASET, H5E_CANTCOPY, FAIL, "unable to copy dataset's datatype");
-            if (NULL == (dst_type = H5T_copy(copied_fill.type, H5T_COPY_TRANSIENT)))
```

Comment:
Any copy call that used `H5T_COPY_ALL` was able to be removed, but the `H5T_COPY_TRANSIENT` ones were left in for now as the code seems to want to set the location of the datatype such that it's a memory type. Since this is only needed for variable-length and reference types, I updated this to only make transient copies of the type if it's one of those two classes. We can probably optimize this further by pushing the copy calls down into the conversion routine and caching those copies.